src/metabase/api/collection.clj

```diff
@@ -102,6 +102,32 @@
           (dissoc ::collection.root/is-root?)
           collection/personal-collection-with-ui-details))))
 
+(defn- shallow-tree-from-collection-id
+  "Returns only a shallow Collection in the provided collection-id, e.g.
+
+  location: /1/
+  ```
+  [{:name     \"A\"
+    :location \"/1/\"
+    :children 1}
+    ...
+    {:name     \"H\"
+     :location \"/1/\"}]
+
+  If the collection-id is nil, then we default to the root collection.
+  ```"
+  [colls]
+  (->> colls
+       (map collection/personal-collection-with-ui-details)
+       (collection/collections->tree nil)
+       (map (fn [coll] (update coll :children #(boolean (seq %)))))))
+
+(defn- location-from-collection-id-clause
+  [collection-id]
+  (if collection-id
+    [:like :location (tru "%/{0}/%" collection-id)]
```

Comment:
I'm afraid this still has a similar issue of reading _all_ the children of the desired collection from the DB, even though only the direct children are returned.

It's a little opaque, but I think you can do what you need with something like 
```suggestion
    [:and
     [:like :location (format "%%/%d/%%" collection-id)]
     [:not [:like :location (format "%%/%d/%%/%%/%%" collection-id)]]]
```

Essentially, you're selecting all collections which have `collection-id` in their parent path, but not ones where it is succeeded by two or more slashes. So you're limiting the results to only children and sub-children of `collection-id`.

There's probably other ways to do it too with subqueries or maybe regexes (although that might require different queries for each app DB) but this was the simplest thing I could think of. I also changed `tru` to `format` in the suggestion code — don't use `tru` for strings that don't actually need to be translated, or else they'll end up in our i18n bundle. A plain `str` could work here too if the doubled-up `%%` are too annoying to read (that's how they're escaped in format strings).

Reply:
It would also be great if you could add a unit test for the query itself, to ensure it's only reading the necessary collections _prior_ to all the post-processing of the tree. Might need to pull the query out as a helper function, but that seems like a good thing.

Reply:
Yea that makes sense, thanks for the info on `tru`!